Pep480 update: adapt to sigstore (#1)
axelsimon/peps: pep-0480.txt

Last-Modified: $Date$
Author: Trishank Karthik Kuppusamy <karthik@trishank.com>,
        Vladimir Diaz <vladimir.diaz@nyu.edu>,
        Justin Cappos <jcappos@nyu.edu>, Marina Moore <mm9693@nyu.edu>,
	axel simon <axel@redhat.com>
BDFL-Delegate: Donald Stufft <donald@stufft.io>
Discussions-To: Packaging category on Python Discourse <https://discuss.python.org/c/packaging>
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Requires: 458
Created: 08-Oct-2014

Abstract
========

Proposed is an extension to PEP 458 that adds support for the "maximum security

> model" described therein: distributions signed by developers. End-to-end

Reviewer: I'm not sure I like "end-to-end" in this PEP anymore, especially because the "end" can be ambiguous. Do mean PyPI? Do we mean the CI/CD? Do we mean developers? But that is a larger issue not relevant to this PR, and we can shelve it for another time.

signing allows both PyPI and developers to sign for the distributions that are
downloaded by clients. The minimum security model proposed by PEP 458 supports
continuous delivery of distributions (because they are signed by online keys),
but that model does not protect distributions in the event that PyPI is
compromised. In the minimum security model, attackers who have compromised the

> signing keys stored on PyPI Infrastructure may sign for malicious

Reviewer: ```suggestion
online signing keys stored on PyPI Infrastructure may sign for malicious
```

distributions. The maximum security model, described in this PEP, retains the
benefits of PEP 458 (e.g., immediate availability of distributions that are

> uploaded to PyPI), but additionally ensures that end-users are not at risk of

Reviewer: ```suggestion
uploaded to PyPI), but additionally ensures that, should PyPI be compromised, end-users of projects delegated to developers before the compromise are not at risk of
```

> installing forged software if PyPI is compromised, so long as developer

Reviewer: ```suggestion
installing forged software, so long as developer
```

> signatures are verified.

Reviewer: ```suggestion
signatures, if any, are verified.
```

This PEP requires some changes to the PyPI infrastructure, and some suggested
changes for developers who wish to participate in end-to-end signing. These
changes include updating the metadata layout from PEP 458 to include delegations
to developer keys, adding a process to register developer keys with PyPI, and a
change in the upload workflow for developers who take advantage of end-to-end
signing. All of these changes are described in detail later in this PEP. Package

> managers who wish to take advantage of end-to-end signing do not need to do any

Reviewer: Please drop the "who": package managers are not humans.

additional work beyond what is required to consume metadata described in PEP

> 458. This PEP further advocates for high-priority and newly created packages to

Reviewer 2: We should define the term "high-priority" explicitly, including a recommendation for how to calculate what is a high-priority package. See section 7.2 of @trishankatdatadog's [Diplomat](https://theupdateframework.io/papers/protect-community-repositories-nsdi2016.pdf?raw=true) paper.

Reviewer: Yeah, this addition seems unnecessarily prescriptive. It might be difficult to get all new projects to start signing until we get the UX almost entirely transparent. Please see the Diplomat paper for policies that we found most effective at the time.

be signed for by developers (rather than all packages or making signing
optional).

This PEP discusses the changes made to PEP 458 but excludes its informational
elements to primarily focus on the maximum security model. For example, an
overview of The Update Framework or the basic mechanisms in PEP 458 are not
covered here. The changes to PEP 458 include modifications to the snapshot
process, key compromise analysis, auditing snapshots, and the steps that should
be taken in the event of a PyPI compromise. The signing and key management

> process that PyPI MAY RECOMMEND is discussed but not strictly defined.

Reviewer: Isn't this not true anymore with this PR? Aren't we prescribing how developers should use Fulcio?

Ultimately, how the release process should be implemented to manage keys and
metadata is left to the implementors of the signing tools. That is, this PEP
delineates the expected cryptographic key type and signature format included in
metadata that MUST be uploaded by developers in order to support end-to-end
verification of distributions.

PEP Status
==========

The community discussed this PEP from 2014 to 2018. Due to the amount of work
required to implement this PEP, discussion was deferred until the precursor
step in PEP 458 was approved. As of mid-2020 PEP 458 is approved and

> implementation is in progress, and the PEP authors aim to gain approval so they

Reviewer 2: ```suggestion
implementation is in progress, and the PEP authors aimed to gain approval so they
```
Past tense?

can secure appropriate funding for implementation. As of the second half of
2021, work is restarted on this PEP to make it ready for implementation once
the changes needed in 458 have landed.

Rationale
=========

PEP 458 [1]_ proposes how PyPI should be integrated with The Update Framework
(TUF) [2]_. It explains how modern package managers like pip can be made more
secure, and the types of attacks that can be prevented if PyPI is modified on
the server side to include TUF metadata. Package managers can reference the
TUF metadata available on PyPI to download distributions more securely.

PEP 458 also describes the metadata layout of the PyPI repository and employs
the minimum security model, which supports continuous delivery of projects and
uses online cryptographic keys (stored and used on the PyPI servers) to sign
the distributions uploaded by developers. Although the minimum security model
guards against most attacks on software updaters [5]_ [7]_, such as
mix-and-match and extraneous dependencies attacks, it can be improved to

> support end-to-end signing and to prohibit forged distributions in the event

Reviewer: ```suggestion
support end-to-end signing and to prohibit forged distributions, in the event
```

> that PyPI is compromised by adding a second layer of signing.

Reviewer: ```suggestion
that PyPI is compromised, by adding a second layer of signing.
```

The threat model assumes the following:

> * Developer keys are either offline or ephemeral. Offline keys are safe and

Reviewer 3: This is a fairly strong assumption.  Surely at least some such keys will be able to be controlled by an attacker.  We don't want all projects on the repo to be insecure if one project has its keys compromised.

Author: "Offline keys are safe and securely stored" was in the PEP when i first read it.
I added the option for ephemeral keys, which should be only written to RAM and never to disk and then flushed after use, usually within a couple minutes of creation.
Is your concern with the overall assumption (offline + ephemeral keys are safe) or the ephemeral part specifically?

Reviewer 2: I think the part of this change being objected to is "Offline keys are safe and securely stored.". The threat model should assume that some offline keys will be able to be controlled by an attacker, and that a combination of thresholds (especially for top-level metadata) and TUF's compromise resilience features (delegations, easy revocation) will protect against offline keys being attacker controlled.

Reviewer: I think I see the confusion here. We need a few bullet points, like so, really:

> * Offline keys exist for a long time, and are safe and securely stored.
> * Ephemeral keys only exist for the time needed to sign and disposed of immediately after.
> * Some critical PyPI keys are offline, while the rest are online.
> * Some developer keys are either offline or ephemeral.

Definitions
===========

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
document are to be interpreted as described in `BCP 14`__ [`RFC2119`__]
[`RFC8174`__] when, and only when, they appear in all capitals, as shown here.
__ https://www.rfc-editor.org/info/bcp14
__ https://www.ietf.org/rfc/rfc2119.txt
__ https://datatracker.ietf.org/doc/html/rfc8174

> This PEP focuses on integrating TUF with PyPI; however, the reader is

Reviewer 2: ```suggestion
This PEP focuses on extending the TUF integration with PyPI; however, the reader is
```
to emphasise that this is additive, building on top of PEP 458

* Roles: Roles in TUF encompass the set of actions a party is authorized to
  perform, including what metadata they may sign and which packages they are
  responsible for. There is one *root* role in PyPI. The *root* role delegates

>   responsibilities to multple roles, directly or indirectly. The term "top-level

Reviewer 4: ```suggestion
  responsibilities to multiple roles, directly or indirectly. The term "top-level
```

>   role" refers to the *root* role and any role delegated by the *root* role.

Reviewer 4: ```suggestion
  role" refers to the *root* role and any role delegated directly by the *root* role.
```

>   role" refers to the *root* role and any role delegated directly by the *root* role.

Reviewer 2: should we list the TUF top-level roles explicitly here (root, timestamp, snapshot, top-level targets)?

  project.

* Online key: A private cryptographic key that MUST be stored on the PyPI
  server infrastructure. This usually allows automated signing with the key.
  An attacker who compromises the PyPI infrastructure will be able to
  immediately read these keys.

* Offline key: A private cryptographic key that MUST be stored independently of
  the PyPI server infrastructure. This prevents automated signing with the
  key. An attacker who compromises the PyPI infrastructure will not be able to

>   immediately read these keys.

Reviewer: We should also define ephemeral keys after this.

* Threshold signature scheme: A role can increase its resilience to key
  compromises by specifying that at least t out of n keys are REQUIRED to sign
  its metadata. A compromise of t-1 keys is insufficient to compromise the
  role itself. Saying that a role requires (t, n) keys denotes the threshold
  signature property.

* OpenID Connect identity: An identity backed by an OpenID Connect provider,
  such as an email address or website account. Examples of these might be a
  Gmail email address, a GitHub account, or any organisation-managed account
  compatible with the OIDC standard and protocol.

> * sigstore: An open source project and associated tools, backed by a public

Reviewer 2: nit: it's confusing that sigstore is the tools and the public good instance, but for the purpose of this PEP I think it's a public good instance backed by an open source project and associated tools?

Reviewer: Not to mention that sigstore is many things, not just one, and people tend to get confused. For the purposes of this PEP, we are interested only in Rekor, Fulcio, and the timestamping (RFC 3161) server.

  good service designed for simplified software artefact signing. It allows for

>   the use of both long-lived and ephemeral cryptographic keys (deleted after

Reviewer 2: Perhaps add an explicit term definition for ephemeral keys, rather than defining it inline?

  signing) and provides a certificate authority combined with a signature log to
  record the keys used to create signatures, allowing for later verification,
  after the keys have been deleted in the case of ephemeral keys.

Maximum Security Model
======================

The maximum security model permits developers to sign their projects and to
upload signed metadata to PyPI. In the model proposed in this PEP, if the PyPI
infrastructure were compromised, attackers would be unable to serve malicious
versions of a *claimed* project without also having access to that project's
developer key. Figure 1 depicts the changes made to the metadata layout of the
minimum security model, namely that developer roles are now supported and that
three new delegated roles exist: *claimed*, *recently-claimed*, and
*unclaimed*. The *bins* role from the minimum security model has been renamed
*unclaimed* and can contain any projects that have not been added to *claimed*.

> The *unclaimed* role functions just as before (i.e., as explained in PEP 458,

Reviewer 2: ```suggestion
The *unclaimed* role functions just as before (i.e., the *bins* role as explained in PEP 458,
```

projects added to this role are signed by PyPI with an online key). Offline
keys provided by developers ensure the strength of the maximum security model
over the minimum model. Although the minimum security model supports continuous
delivery of projects, all projects are signed by an online key. That is, a
successful attacker of the PyPI infrastructure is able to corrupt packages in
the minimum security model, but not in the maximum model, without also

> compromising a developer's key.

Reviewer: Nitpick: going forward, could we please separate strictly formatting/linguistic changes from the meat (e.g., Rekor/Fulcio)? It would made reading PRs much easier. Thanks!

==================

End-to-end signing allows both PyPI and developers to sign for the metadata
downloaded by clients. PyPI is trusted to make uploaded projects available to
clients (PyPI signs for the metadata for this part of the process), and
developers sign for the distributions that they upload to PyPI.

> In order to delegate trust to a project, developers are required to submit an

Reviewer: I disagree with the "required" bit. As someone else (@joshuagl?) mentioned elsewhere in this PR, we should give developers as many options as possible for managing their keys. Sometimes, it can be with Fulcio, and sometimes it can be with GPG, and so on.

> OpenID Connect identity or a public key to PyPI. PyPI takes the project's OIDC

Reviewer 3: Forgive my ignorance, but can you say more about how PyPI would delegate from a targets role using an OIDC identity...

Reviewer 5: I agree, some more detail here would be helpful. I think we could paraphrase a bit from https://docs.sigstore.dev/cosign/openid_signing/ a bit, and describe both the OAuth flow, and the identity token flow in an automated environment.

> identity or public key and adds it to parent metadata, which PyPI then signs.

Reviewer 5: Not sure I follow what's being suggested here.

Reviewer 6: Ditto -- is the OIDC lifecycle specified here intended to be different from the one we're developing for credential-less publication? My understanding (which could be very wrong!) was that we were planning to use that one lifecycle for both.

For context, here's the authentication lifecycle as we're currently planning it: https://github.com/pypa/warehouse/issues/10619

Reviewer 2: @axelsimon is this trying to describe a TUF role on PyPI (*claimed*) delegating to a developer key/identity??

Reviewer: What is the project's OIDC "identity" vs the "public key"? This is totally unclear to me right now.

After the initial trust is established, developers are required to sign
distributions that they upload to PyPI using an ephemeral sigstore key, recorded
in a certificate in sigstore's certificate authority, or should they so desire,

> with the private key corresponding to their public key. The signed TUF metadata

Reviewer 2: This feels awkward and I see similar patterns throughout the edit. Can we define developer signing once as being either OIDC identity-backed ephemeral keys or long-lived (ideally hardware token based) keys and then just use the term "developer sign[ed|ing]" throughout the PEP?

Reviewer: Ditto. Also, developers should not be required to use Fulcio.

that developers upload to PyPI includes information like the distribution's file
size and hash, which package managers use to verify distributions that are
downloaded.

> Two naive approaches regarding end-to-end signing exist: require signing of any

Reviewer: Understand this concern, but would remove it from this PR, as it adds an unnecessary prescription. This PR should be solely about the optional use of Fulcio for managing developer keys.

and all new distribution uploads, or make signing entirely voluntary, making it
likely to render signing an unused feature.
This PEP advocates for a third approach: require high-priority and newly created
packages to be signed for by developers.

The practical implications of end-to-end signing is the extra administrative
work needed to delegate trust to a project, and the signed metadata that
developers MUST upload to PyPI along with the distribution. Specifically, PyPI
is expected to periodically sign metadata with an offline key by adding
projects to the *claimed* metadata file and signing it. In contrast, in the

> minimum ssecurity model, projects are only ever signed with an online key.

Reviewer 5: ```suggestion
minimum security model, projects are only ever signed with an online key.
```

Author: Oops. Thanks, good catch.

passwords with PBKDF2-HMAC-SHA256 (100K iterations by default, but this may be
overridden by the developer). The current Python implementation of TUF can use
any cryptographic library (support for PyCA cryptography will be added in the
future), may override the default number of PBKDF2 iterations, and the KDF (Key

> Derivation Funciton) tweaked to taste.

Reviewer 2: ```suggestion
Derivation Function) tweaked to taste.
```

__ https://en.wikipedia.org/wiki/Transparency_%28human%E2%80%93computer_interaction%29

> Therefore, an easy-to-use signing solution is RECOMMENDED for developers. One

Reviewer: Precisely! Recommended, not required.

solution is to use ephemeral keys, where the private key is only kept in memory
(and not stored on disk) for as long as needed to produce the signature and is
then safely discarded. This removes any long-term challenge of protecting the
key by doing away with the key in the first place.

`Sigstore`__ provides a way to create and use such ephemeral keys and maintain
a chain of trust between the signature made on a distribution and the publicly
available information stored on its transparency log. It uses certificates

> issued by ts certificate authority (Fulcio) to prove a signing key was valid

Reviewer: TS?

Reviewer 7: Looks like a typo for 'its' to me.

when a signature was made. In-depth discussion of sigstore's architecture is
beyond the scope of this document and for more information one may refer to

> https://sigstore.dev

Reviewer: This is the Appendix, and I believe we should go into enough details about how the proposed solution would work here, the same way we described how TUF works in PEP 458.

Ephemeral keys tie to an OpenID Connect identity, which is portable across
multiple computers and does not require any key sharing. It also relies on an
account a developer already uses, removing the issue of requiring a further
password to remember.

Optionally, sigstore also allows for manually managed keypairs. Its cosign tool

> can generate keys using the Ed25519 signature scheme.

Reviewer 2: ```suggestion
can generate keys using the Ed25519 signature scheme or use keys from existing
hardware security modules (HSM) such as a YubiKey.
```
```suggestion
can generate keys using the Ed25519 signature scheme.
```

__ https://github.com/sigstore/

The following outlines an automated signing solution that a new developer MAY
follow to upload a distribution to PyPI:

1.  Register a PyPI project.
2.  Register an OIDC identity for the project.

>   2.1 Alternative: register a public key.

Reviewer 3: Can this be organized differently so that registering a public key and an OIDC identity are at the same level?  It seems more natural that way.

Reviewer 5: IMO, PyPI is far less likely to support public keys than ephemeral keys via Sigstore. I think this PEP should probably just make the best possible recommendation, and avoid providing alternatives.

Reviewer 2: Based on discussions around sigstore I've observed, I'd expect some pushback on _only_ supporting sigstore ephemeral keys. Those objections (distrust of valid OIDC providers, unwillingness to leak identity, perfectly happy with personal GPG key) would perhaps be somewhat mitigated by PyPI becoming an OIDC provider, though that has the consequence of making PyPI a larger target and single point of failure.

Reviewer: Yes, I'm in the camp that we should support alternatives, but at the same time recommend only one of them. If we're willing to go as far as running an OIDC server on PyPI, it doesn't seem more difficult to add support for offline, manually-managed developer keys, as the key distribution and verification is actually simpler.

  2.2 Optional: Add a new identity to the developer's PyPI user account from a
    second machine.
3.  Upload project.

Step 1 is the normal procedure followed by developers to `register a PyPI
project`__.

__ https://pypi.python.org/pypi?:action=register_form

> Step 2 registers an OIDC identity to PyPI, generates an ephemeral key and signs the

Reviewer: I think we should rewrite this: the step doesn't register anything --- developers do.

TUF metadata that is generated for the distribution.

> Optionally, a long-lived keypair MAY be generated instead (where the private

Reviewer: Agree with Justin and Dustin: we should talk about one option at a time, not both.

key file SHOULD be encrypted) and the Ed25519 public key uploaded to PyPI.

Also optionally, adding a new identity from a second machine in step 2.2, can
be done in two ways. First, by adding a second OIDC identity to the PyPI

> project. Second, by creating a second keypair, which also generates an

Reviewer 2: This is confusing and feels out of scope? If you want to sign from a second machine operated by the same developer, wouldn't you just use the same OIDC identity?

encrypted private key file and uploads an Ed25519 public key to PyPI.

Separate identities MAY be created to allow a developer, or other project
maintainers, to sign releases on multiple machines. An existing verified
identity (its public key is contained in project metadata or has been uploaded
to PyPI) signs for new identities. By default, project metadata has a signature
threshold of "1" and other verified identities may create new releases to
satisfy the threshold. This second solution puts a stronger burden on

> developers and is not recommended.

Reviewer 2: Are we recommending against threshold signing here? I don't think we should do that. It seems valuable to have a threshold of maintainers sign for a release where possible.
